### PR TITLE
Remove the use of dmlc object registry

### PR DESCRIPTION
**Pull request by hcho3**
Extracted from #285.

* Do not use the header `dmlc/registry.h`.
* Do not use the object registry provided by dmlc-core. The registry lets you instantiate a new object by the class name. We really don't need the registry, since there are only two kinds of code generators (`FailSafeCompiler` and `ASTNativeCompiler`). Simple `if` / `else` statements suffice.
* Use pimpl pattern to hide templated functions in the code generators.
* Define new headers `failsafe.h` and `ast_native.h` to store the class declarations of `FailSafeCompiler` and `ASTNativeCompiler`. This is more in line with conventional C++.
* Rename `param` to `param_`.

### CODE REVIEW

File: src/compiler/failsafe.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+class FailSafeCompilerImpl;
+
+class FailSafeCompiler : public Compiler {
+ public:
+  explicit FailSafeCompiler(const CompilerParam& param);
+  CompiledModel Compile(const Model& model) override;
+ private:
+  std::unique_ptr<FailSafeCompilerImpl> pimpl_;
+};
```
Comment (trivialfis) [inline src/compiler/failsafe.h:29]: Interesting, I always find using unique ptr with incomplete class troublesome due to missing destructor.  Not sure how did you get it to work.

File: src/compiler/failsafe.cc [hunk 2]
```diff
@@ -389,17 +367,40 @@ class FailSafeCompiler : public Compiler {
...
 
-TREELITE_REGISTER_COMPILER(FailSafeCompiler, "failsafe")
-.describe("Simple compiler to express trees as a tight for-loop")
-.set_body([](const CompilerParam& param) -> Compiler* {
-    return new FailSafeCompiler(param);
-  });
+FailSafeCompiler::FailSafeCompiler(const CompilerParam& param)
+    : pimpl_(std::make_unique<FailSafeCompilerImpl>(param)) {
+  if (param.verbose > 0) {
+    LOG(INFO) << "Using FailSafeCompiler";
```
Comment (trivialfis) [inline src/compiler/failsafe.cc:400]: Is there a log warning macro?

Comment (hcho3) [reply src/compiler/failsafe.cc:400]: LOG(WARN) is just an alias of LOG(INFO). Treelite doesn't let you set global verbosity level like XGBoost.

File: src/compiler/failsafe.h (hunk 1 above)
Comment (hcho3) [reply src/compiler/failsafe.h:29]: Take a look at https://www.fluentcpp.com/2017/09/22/make-pimpl-using-unique_ptr/. The key is to explicitly define the destructor for `FailSafeCompiler` after the full definition of `FailSafeCompilerImpl` is provided. See my latest commit.

File: src/compiler/failsafe.cc (hunk 2 above)
Comment (trivialfis) [reply src/compiler/failsafe.cc:400]: Got it.

File: src/compiler/failsafe.h (hunk 1 above)
Comment (trivialfis) [reply src/compiler/failsafe.h:29]: Thanks for the reference.  That's a weird aspect of c++.

**Review by trivialfis — APPROVED** [review]

**Pull request merged**